### DISCUSSION TITLE
Is there anyway to customize the path to the build meta json files?

### DISCUSSION THREAD

**Discussion by ahku** [nuxt/nuxt#25448] [Questions]
Hi, I'm trying to launch my Nuxt application as a block in a CMS, and it has not been easy. I have to upload all the assets to a file manager, and update the HTML with the hashed assets links. Could not figure out a way to remove the hash, but it's okay for now. 

What's not okay is that the app needs to load some json files in /builds/meta, and the only way I've been able to make it load from the CDN path, is to modify function that wraps the path in the entry.xxx.js file:

![CleanShot 2024-01-26 at 14 19 09@2x](https://github.com/nuxt/nuxt/assets/1736890/6250a019-1b31-45ef-9a7d-08d2c3683555)

Has anybody done something similar, or have any idea how to best solve this?

**Comment by das-nirmal** [comment]
Hi @ahku I faced similar issue - planning to resolve by setting the experimental.appManifest in nuxt.config to false. 

Refer 
https://nuxt.com/blog/v3-8#app-manifest
https://github.com/nuxt/nuxt/pull/21641

**Comment by beltreenmanuel908-afk** [comment]
https://nuxt.com/blog/v3-8#app-manifest

**Comment by ORCHORDS** [comment]
The path is hardcoded; there is no config-level option to move it without forking Nuxt. Verified directly from the Nuxt 3.8 blog post that introduced the feature:

> "We now support a built-in app manifest (see PR #21641), which generates a manifest at `/_nuxt/builds/meta/<buildId>.json`."

And the experimental-features guide for Nuxt 4:

> "Use app manifests to respect route rules on client-side. This flag is enabled by default, but you can disable this feature: experimental: { appManifest: false }"

The blog post explicitly notes only the disable flag; no path-customization flag is documented. So your three real options, in order of preference:

1. Disable the manifest if you do not need it. The manifest is a build-id metadata file used by route rules on the client. If your app does not use route rules or does not need the build-id mismatch detection on the client, set experimental.appManifest: false in nuxt.config.ts. The /builds/meta/ folder simply will not be generated. This is the option das-nirmal and beltreenmanuel908-afk are pointing you at, and it is the right answer for the "embedded in a CMS" use case where you are uploading pre-built assets manually and want stable paths.

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  experimental: {
    appManifest: false,
  },
})
```

2. If you must keep the manifest but want a stable URL, the buildId component is the only path segment that varies per build. You can keep /builds/meta/ as a fixed folder and accept that the file inside changes its name with each deploy. To do this:

   - Keep experimental.appManifest: true (default).
   - On deploy, your CMS-side orchestrator fetches /_nuxt/builds/meta/*.json (or scans the directory if the CMS supports it) and uploads the latest one to whatever fixed URL your CMS expects, e.g. /builds/meta/manifest.json.
   - On the client, point your app at /builds/meta/manifest.json instead of the hashed URL.

   This is what most CMS-embedded Nuxt apps end up doing because the manifest is small (a few hundred bytes of JSON) and the rebuild cost on deploy is low.

3. Patch the path via a small Nitro plugin that intercepts the request and rewrites it. This is the workaround you found (modifying entry.xxx.js). A cleaner version lives in nuxt.config.ts as a Nitro hook:

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  hooks: {
    'nitro:config'(nitroConfig) {
      // Rewrite /_nuxt/builds/meta/*.json to /builds/meta/manifest.json
      // for any incoming GET request.
      nitroConfig.handlers = nitroConfig.handlers || []
      nitroConfig.handlers.unshift({
        method: 'get',
        middleware: false,
        handler: '~/server/middleware/app-manifest-rewrite.ts',
      })
    },
  },
})
```

And then server/middleware/app-manifest-rewrite.ts:

```ts
// server/middleware/app-manifest-rewrite.ts
export default defineEventHandler((event) => {
  const url = getRequestURL(event)
  if (url.pathname.startsWith('/_nuxt/builds/meta/')) {
    const fileName = url.pathname.split('/').pop()
    return sendRedirect(event, `/builds/meta/manifest.json?v=${fileName?.replace('.json', '')}`, 302)
  }
})
```

The query string preserves the buildId so cache invalidation still works on the CMS side. If your CMS does not handle 302s, switch the response to a stream that reads the actual file from the public folder.

Why option 1 is usually the right answer for embedded apps: the manifest is only useful when the client needs to detect that the deployed server bundle does not match the one the static page expects (which matters for route rules with cache: false and similar). In a CMS-embedded scenario you redeploy the whole thing atomically, so the mismatch never happens, and the manifest is just dead weight.

Cross-references:

- https://nuxt.com/blog/v3-8 (announces the feature, default path, opt-out flag)
- https://nuxt.com/docs/4.x/guide/going-further/experimental-features (current docs, current default, disable snippet)
- PR #21641 (the original implementation)
